### PR TITLE
Add detailed logs for a received, transmitted, or dropped IPv6 message

### PR DESCRIPTION
**Pull request by abtink**
This commit adds detailed logs (at `info` level) when an IPv6
message is received, transmitted or dropped (due to tx failures).
The log contains the following information about the IPv6 message:

- The message length (in bytes),
- The protocol (TCP/UDP/ICMP6, etc.),
- Checksum value for UDP or TCP,
- MAC address of source/dest (short or extended address),
- Message security (enabled/disabled),
- Message's source and destination IPv6 addresses,
- In case of tx failure, error causing the drop (CCA, No Ack, ...).

### CODE REVIEW

File: src/core/thread/mesh_forwarder.cpp [hunk 1]
```diff
@@ -2271,4 +2279,75 @@ void MeshForwarder::HandleDataPollTimeout(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_LOG_LEVEL >= OPENTHREAD_LOG_LEVEL_INFO
```
Comment (nibanks) [inline src/core/thread/mesh_forwarder.cpp]: Why INFO level?

File: src/core/thread/mesh_forwarder.cpp [hunk 2]
```diff
@@ -2271,4 +2279,75 @@ void MeshForwarder::HandleDataPollTimeout(void)
...
+        checksum,
+        (aAction == kMessageReceive) ? "from" : "to",
+        aMacAddress.ToString(stringBuffer, sizeof(stringBuffer)),
+        aMessage.IsLinkSecurityEnabled() ? "yes" : "no",
+        (aError == kThreadError_None) ? "" : ", error:",
+        (aError == kThreadError_None) ? "" : otThreadErrorToString(aError)
+    );
+
+    otLogInfoMac(GetInstance(), "src: %s", ip6Header.GetSource().ToString(stringBuffer, sizeof(stringBuffer)));
+    otLogInfoMac(GetInstance(), "dst: %s", ip6Header.GetDestination().ToString(stringBuffer, sizeof(stringBuffer)));
```
Comment (nibanks) [inline src/core/thread/mesh_forwarder.cpp]: Why don't you combine these into a single log?

File: src/core/thread/mesh_forwarder.cpp (hunk 1 above)
Comment (abtink) [inline src/core/thread/mesh_forwarder.cpp]: `Info` makes sense for this. It's not a warning or error and we want this at a level higher than debug. What would you suggest for it to be?

File: src/core/thread/mesh_forwarder.cpp (hunk 2 above)
Comment (abtink) [inline src/core/thread/mesh_forwarder.cpp]: It makes the log message too long. 
Underlying `otPlatLog()` on a resource-limited platforms typically limit the length of a single log message and clip anything after some point...

**Comment by abtink** [comment]
@nibanks , can you please help me with this windows/appvoyer build error:
```
..\..\src\core\mac\mac_frame.cpp(56): error C3861: 'snprintf': identifier not found [C:\projects\openthread\etc\visual-studio\libopenthread_k.vcxproj]
```
Is there a trick to get `snprint()`? Thanks.

**Comment by nibanks** [comment]
Looks like you are missing an include for `#include <stdio.h>`.

File: src/core/thread/mesh_forwarder.cpp (hunk 1 above)
Comment (nibanks) [inline src/core/thread/mesh_forwarder.cpp]: It's just that on Windows, lots of traffic can get generated, and I'm not sure we want to log every packet if just INFO level is enabled. I think it is fine for now. If it really becomes an issue later, we can work something out.

File: src/core/thread/mesh_forwarder.cpp (hunk 2 above)
Comment (nibanks) [inline src/core/thread/mesh_forwarder.cpp]: My main concern is that if you split related info between multiple logs, on Windows, the information can get easily separated because of multiple threads logging simultaneously. If there is a platform limitation on the length of one log message, should it be the platform's responsibility to split into multiple messages?

**Comment by abtink** [comment]
We have `#include <stdio.h>` in `mac_frame.cpp`  [here](https://github.com/openthread/openthread/pull/1515/files#diff-0052ced0c2f6e36594f4a30bc509429bR34) do we need it somewhere else?

File: src/core/thread/mesh_forwarder.cpp (hunk 1 above)
Comment (abtink) [inline src/core/thread/mesh_forwarder.cpp]: I agree. We may need to later consider adding a new log level (like `notice` in between `info` and `warning`)...or another option is to add a `CONFIG' option to enable/disable logs per IP message.

**Comment by nibanks** [comment]
Then you probably need to include the config file. Take a look at link_quality.cpp.

**Comment by abtink** [comment]
Ok. Thanks... Just pushed a new commit and added the config file as well...

File: src/core/thread/mesh_forwarder.cpp (hunk 2 above)
Comment (abtink) [inline src/core/thread/mesh_forwarder.cpp]: It will not be easy to handle it at platform. Even syslog imposes a log length limit.

About logs on a multi-threaded system, we faced similar case when running multiple instances of wpantund on the same machine (IP message logs in wpantund also span multiple lines). The best approach is to have the thread info (or maybe the OT instance info) appended at the beginning of the log message to be able to distinguish/group logs corresponding to same instance.

Comment (jwhui) [inline src/core/thread/mesh_forwarder.cpp]: I believe we have a similar issue with `otLogDump()` as well.  If we need to address the "multiline logging issue", then I think we should do that more generally in a separate PR.
